### PR TITLE
feature: add new survey notification to design page

### LINKED ISSUE #8029: HaTs telemetry items
- [x] Mechanism for displaying a banner/notification in Composer to show to the users a prompt to complete a survey

- [x] Ability to dynamically generate the survey link parameters populated with values from composer usage (for example Version) & (survey link itself would be stable and hardcoded)

- [x] Ability to store some of the characteristics required for the trigger mechanism in local composer storage&nbsp;(e.g.&nbsp;count of sessions)

### PR DESCRIPTION
**Pull request by beyackle**
## Description

[This is a redo of #8078, which was closed by mistake]

This adds the logic to show a notification on the home screen prompting the user to take a survey. This also adds some new notification types and a fairly generic way to track a user's usage of various features inside the browser's own local storage, which will be useful for other possible business cases (like extending the conditions surveys pop up in, or cosmetic things like Achievements).

## Task Item

refs #8029
[will close when this merges into main]

## Screenshots

image

### CODE REVIEW

File: Composer/packages/electron-server/src/main.ts [hunk 1]
```diff
@@ -295,11 +295,16 @@ async function run() {
...
     const mainWindow = getMainWindow();
+    const machineId = await getMachineId();
+
     mainWindow?.webContents.send('session-update', 'session-started');
 
     if (process.env.COMPOSER_DEV_TOOLS) {
       mainWindow?.webContents.openDevTools();
     }
+
+    log(`Machine ID is ${machineId}`);
```
Comment (srinaath) [inline Composer/packages/electron-server/src/main.ts:306]: Is this log required?

File: Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+const buildUrl = (info: MachineInfo) => {
+  // User OS
+  // hashed machineId
+  // composer version
+  // maybe include subscription ID; wait for global sign-in feature
+  // session ID  (global telemetry GUID)
+  const version = process.env.COMPOSER_VERSION;
+
+  const parameters = {
+    Source: 'Composer',
```
Comment (srinaath) [inline Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx:23]: camelCase why capitalized?

File: Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    os: info?.os || 'Unknown',
+    machineId: info?.id,
+    version,
+  };
+
+  return (
+    `${SURVEY_URL_BASE}?` +
+    Object.keys(parameters)
```
Comment (srinaath) [inline Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx:31]: nit: can use `query-string` stringify method

File: Composer/packages/client/src/App.tsx [hunk 4]
```diff
@@ -42,6 +43,10 @@ export const App: React.FC = () => {
     ipcRenderer?.on('cleanup', (_event) => {
       performAppCleanupOnQuit();
     });
+
+    ipcRenderer?.on('machine-info', (_event, info) => {
+      setMachineInfo(info);
```
Comment (srinaath) [inline Composer/packages/client/src/App.tsx:48]: It looks like you are using the machine info only inside the useSurveyNotification hook. If so, why do we need to create a recoil state for it. You can just use the `ipcRenderer.on` inside the hook. Ex `useInitializeLoger.ts`

**Review by srinaath — CHANGES_REQUESTED** [review]

File: Composer/packages/electron-server/src/main.ts (hunk 1 above)
Comment (beyackle) [reply Composer/packages/electron-server/src/main.ts:306]: It's not required, but it's useful - this goes into the debug log which users won't see unless they're running via the console anyway.

File: Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx (hunk 2 above)
Comment (beyackle) [reply Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx:23]: I know these are usually lowercase, but uppercase S is the spec I was given.

File: Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx (hunk 3 above)
Comment (beyackle) [reply Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx:31]: Good idea. I'll fix that.

File: Composer/packages/client/src/App.tsx (hunk 4 above)
Comment (beyackle) [reply Composer/packages/client/src/App.tsx:48]: I was trying to make this work last week, but it runs into annoying race conditions where the message handler isn't initialized before the message gets sent and so nothing gets through. Using the Recoil store lets us put the machine info down somewhere whenever we receive it and then count on it being available when we need it.

File: Composer/packages/client/src/components/Notifications/NotificationCard.tsx [hunk 5]
```diff
@@ -36,7 +37,7 @@ const cardContainer = (show: boolean, ref?: HTMLDivElement | null) => () => {
     border-left: 4px solid #0078d4;
     background: white;
     box-shadow: 0 6.4px 14.4px 0 rgba(0, 0, 0, 0.132), 0 1.2px 3.6px 0 rgba(0, 0, 0, 0.108);
-    min-width: 340px;
+    width: 340px;
```
Comment (tdurnford) [inline Composer/packages/client/src/components/Notifications/NotificationCard.tsx:40]: Why did you change this from `min-width` to `width`?

Comment (tdurnford) [inline Composer/packages/client/src/components/Notifications/NotificationCard.tsx:40]: Why did you change this from `min-width` to `width`?

File: Composer/packages/client/src/App.tsx (hunk 4 above)
Comment (srinaath) [reply Composer/packages/client/src/App.tsx:48]: Makes sense to usethe recoil state unless the hook can be registered in app.tsx

File: Composer/packages/client/src/components/Notifications/NotificationCard.tsx [hunk 6]
```diff
@@ -112,13 +113,20 @@ const cardDescription = css`
...
-const linkButton = css`
-  color: #0078d4;
-  float: right;
-  font-size: 12px;
-  height: auto;
-  margin: 4px 0 4px 8px;
-`;
+const linkButton = {
+  root: {
+    padding: '0',
```
Comment (tdurnford) [inline Composer/packages/client/src/components/Notifications/NotificationCard.tsx:118]: `padding` and `border` should be `0px`. Also, why are you setting the border styles? `ActionButton`s don't have borders by default.

File: Composer/packages/client/src/components/Notifications/NotificationCard.tsx [hunk 7]
```diff
@@ -148,27 +156,58 @@ export type NotificationProps = {
...
-        ))}
+        <Stack horizontal horizontalAlign="space-between">
+          <Stack {...stackProps}>
+            {leftLinkList.map(
+              (link) => link != null && <Stack.Item key={link.label}>{makeLinkLabel(link)}</Stack.Item>
+            )}
+          </Stack>
+          <Stack {...stackProps}>
+            {rightLinkList.map(
+              (link) => link != null && <Stack.Item key={link.label}>{makeLinkLabel(link)}</Stack.Item>
```
Comment (tdurnford) [inline Composer/packages/client/src/components/Notifications/NotificationCard.tsx:207]: Why would a link be null?

File: Composer/packages/electron-server/src/main.ts (hunk 1 above)
Comment (tdurnford) [reply Composer/packages/electron-server/src/main.ts:306]: Logs are only visible from the terminal, so I think it's fine.

File: Composer/packages/types/src/shell.ts [hunk 8]
```diff
@@ -67,12 +67,16 @@ export type DisabledMenuActions = {
...
 
 export type Notification = {
-  type: 'info' | 'warning' | 'error' | 'pending' | 'success' | 'question' | 'congratulation';
+  type: 'info' | 'warning' | 'error' | 'pending' | 'success' | 'question' | 'congratulation' | 'custom';
   title: string;
   description?: string;
   retentionTime?: number;
   link?: NotificationLink;
-  links?: NotificationLink[];
+  links?: (NotificationLink | null | undefined)[];
```
Comment (tdurnford) [inline Composer/packages/types/src/shell.ts:75]: Why are we adding null or undefined? Seems very unnecessary.

File: Composer/packages/client/src/App.tsx (hunk 4 above)
Comment (tdurnford) [reply Composer/packages/client/src/App.tsx:48]: Can you explain the race condition a bit more. I'm also confused why we can't just initialize the survey here when the event is received? If the message handler doesn't get initialized before the message is sent, we would never get the message, right?

**Review by tdurnford — CHANGES_REQUESTED** [review]

File: Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    Source: 'Composer',
+    os: info?.os || 'Unknown',
+    machineId: info?.id,
+    version,
+  };
+
+  return `${SURVEY_URL_BASE}?${querystring.stringify(parameters)}`;
+};
+
+export function useSurveyNotification() {
```
Comment (tdurnford) [inline Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx:33]: Arrow function

File: Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
```
Comment (tdurnford) [inline Composer/packages/client/src/components/Notifications/useSurveyNotification.tsx:3]: This should be a ts instead of a tsx file since it is a hook and it does not have any react components

File: Composer/packages/client/src/components/Notifications/NotificationCard.tsx (hunk 7 above)
Comment (beyackle) [reply Composer/packages/client/src/components/Notifications/NotificationCard.tsx:207]: I was using that to indicate a gap, to make the design work, but you're right that this is unnecessary now. I'll fix it.

File: Composer/packages/types/src/shell.ts (hunk 8 above)
Comment (beyackle) [reply Composer/packages/types/src/shell.ts:75]: It is, now that I've split them into left and right groups.

File: Composer/packages/client/src/components/Notifications/NotificationCard.tsx (hunk 6 above)
Comment (beyackle) [reply Composer/packages/client/src/components/Notifications/NotificationCard.tsx:118]: 0 is equivalent to 0 in any unit; there's no need to specify pixels there. And there was a 1 pixel border there that this eliminates, just to align to the designs better.

File: Composer/packages/client/src/components/Notifications/NotificationCard.tsx (hunk 5 above)
Comment (beyackle) [reply Composer/packages/client/src/components/Notifications/NotificationCard.tsx:40]: This is in order to keep the width of a card consistent when the text inside needs to wrap to multiple lines. They fit into the sidebar that way.
image

File: Composer/packages/electron-server/src/utility/machineId.ts [hunk 11]
```diff
@@ -13,17 +13,20 @@ import { readTextFileSync, writeJsonFileSync } from './fs';
 
 export const persistedFilePath = path.join(app.getPath('userData'), 'persisted.json');
 
+function truncate(str: string): string {
```
Comment (tdurnford) [inline Composer/packages/electron-server/src/utility/machineId.ts:16]: Arrow function

File: Composer/packages/lib/code-editor/src/components/toolbar/__tests__/ToolbarButtonMenu.test.tsx [hunk 12]
```diff
@@ -35,7 +35,7 @@ const templatePayload: TemplateRefPayload = {
 const propertiesPayload: PropertyRefPayload = {
   kind: 'property',
   data: {
```
Comment (tdurnford) [inline Composer/packages/lib/code-editor/src/components/toolbar/__tests__/ToolbarButtonMenu.test.tsx:37]: Why are these tests being changed? They're not related to this work at all.

Comment (beyackle) [reply Composer/packages/lib/code-editor/src/components/toolbar/__tests__/ToolbarButtonMenu.test.tsx:37]: They're being changed because they were failing mysteriously. I feel like the word "this" is too common to be used as a name and search target in tests like this without false positives happening.

File: Composer/packages/electron-server/src/utility/machineId.ts (hunk 11 above)
Comment (beyackle) [reply Composer/packages/electron-server/src/utility/machineId.ts:16]: I don't see what that would get us here. This isn't inside a component that gets re-rendered, this isn't exported anywhere, it's just a simple utility function that belongs to this one file.

Comment (tdurnford) [reply Composer/packages/electron-server/src/utility/machineId.ts:16]: The other functions in this file use arrow functions. Best to be consistent. More of a nit.
